### PR TITLE
Vanilla hacker news Implementation

### PR DESCRIPTION
**Pull request by cristianbote**
Hi there!

I'd like to add my 'vanilla' version for a HN clone, using nothing but plain javascript. For the moment, I did not used any server-side rendering, and I'm using the unofficial api, just because of ease of use.

The scores are not as low as I would like to have them, and I'm blaming mostly the hosting, since on real devices is flying super fast. Noticed this in the timeline trace on wp.

Let me know if I need to tinker with something.

Thank you and cheers!

Edit: The current numbers differs in when you do the wp test. In the morning (my morning UTC+2) I got better numbers, than doing them in the evening. I think it's a combination of the HN api and firebase hosting usages.

### CODE REVIEW

**Comment by cheeaun** [comment]
Nice, love the 🍦 logo 😆

**Comment by cristianbote** [comment]
@cheeaun :) Thanks! Couldn't figure out something else, that stands for vanilla, but it fits nicely.

**Comment by davideast** [comment]
@cristianbote This looks awesome!

Can you do me a favor and try out the new (somewhat experimental) HNPWA API? It's benefit is that it uses the Dynamic Firebase Hosting feature with CDN caching, so the latency is lower worldwide. Hopefully it will help drop your times.

It's hosted at hnpwa.com/api/v0/ and follows the same structure as the unofficial node-api you're using, but just appends everything with `.json`. It should just be a replaced of the base url and the extension.

https://hnpwa.com/api/v0/news.json
https://hnpwa.com/api/v0/ask.json
https://hnpwa.com/api/v0/show.json
https://hnpwa.com/api/v0/jobs.json
https://hnpwa.com/api/v0/item/1.json

Let me know if you're interested!

**Comment by cristianbote** [comment]
Hi @davideast, that's awesome! I have almost -1s improvements for faster 3g!

Although this, looks blazingly fast, wp adds at least 400ms, for EM, at each request state. So, for each ajax requests I have 1.2s penalty. I guess, here's the server-side rendering advantage.
But, this _is_ the test that an pwa should go through, these are the worst conditions that could happen for your user, so it's not real-world accurate, but it's a disaster scenario where the user should be able to use your app.

Adding the latest updates to the report.

Also, any idea why the HSTS redirect is not considered by PWA guidelines, as a ok?

**Comment by davideast** [comment]
@cristianbote This is where we can get really clever :)

First, use H/2 Server Push to push any initial render based CSS/HTML/JS down with the original document. Don't include images though, that may just slow things down. You can do this in your `firebase.json` through Link headers:

```json
{
  "hosting": {
    "headers": [
      {
        "source": "/",
        "headers": [{"key": "Link", "value": "</js/app.js>;rel=preload;as=script,</css/app.css>;rel=preload;as=style"}]
      },
      {
         "source": "/users/*",
        "headers": [{"key": "Link", "value": "</js/app.js>;rel=preload;as=script,</css/app.css>;rel=preload;as=style;nopush,</css/users.css>;rel=preload;as=style"}]
      }
    ]
  }
}
```

Second, we can actually Server Push down the API data. This means when it is requested in your app code it will already be downloaded over the network, thus eliminating that 400ms RT :)

[Check out the official blog post for more info.](https://firebase.googleblog.com/2016/09/http2-comes-to-firebase-hosting.html)

**Comment by addyosmani** [comment]
Just dropping by to say nice work on the implementation, @cristianbote 👍  

I know @davideast has a few ideas around how to improve the perf numbers here even further so I'll hold off on a full review until we've chatted about those some more.

Perhaps the only UX feedback I have on the implementation is it looks great on mobile but desktop, stretching to the width of the viewport makes it a little more work to get the cursor all the way over to the left of the screen. I wonder if it would make sense to adopt responsive margins like some of these implems do:

https://react-hn.kristoferbaxter.com/
https://hn.kristoferbaxter.com/
https://hn.svelte.technology/
https://vue-hn.now.sh/top

**Comment by cristianbote** [comment]
@davideast cool stuff! That worked like a charm :) EM numbers are around 3s, which is so much better.

@addyosmani Thanks for the suggestion! That really made it more visually appealing, so ;) thanks!

If there's anything else, let me know, more than happy to adjust it. Cheers!

**Comment by cristianbote** [comment]
Alrighty, I think I'm done with adjusting the scores. I was going at it, for the last day, and I think just with the hosting configuration for firebase, it's enough. Huge thanks to @davideast for suggestions.

So, I noticed there's been another PR closed ahead of mine, with different weights. Do I have to increment mines now? 😄 

ping @addyosmani

**Comment by addyosmani** [comment]
If you could increment that would definitely help us with landing this sooner :) If another weight gets merged and messes that up we'll clean it up so you don't have to.

**Comment by addyosmani** [comment]
@davideast Could we get your sign-off on this one from the Firebase side?

**Comment by cristianbote** [comment]
@addyosmani done! 😃

**Comment by davideast** [comment]
@addyosmani LGTM!

File: site/_apps/vanilla-hn.txt [hunk 1]
```diff
@@ -1,5 +1,5 @@
 ---
-weight: 16
+weight: 18
```
Comment (addyosmani) [inline site/_apps/vanilla-hn.txt]: I'll bump this to 19 once we merge it in.

**Comment by addyosmani** [comment]
Just going back through the implementation - this one uses a different pagination pattern to the other apps:

image

Whilst the other apps use visible paging which is part of the routing story:

image

e.g https://react-hn.kristoferbaxter.com/top/2

Could we update the UX in this app to follow that pattern? I'm happy to land it after that.

**Comment by cristianbote** [comment]
Hey @addyosmani, indeed I use a different pagination pattern. Shouldn't be a problem updating it, but just one more thing, are we talking just about showing `previous 1/10 next`? or also url changes?
Also, I'm not displaying the index/order of the item, should that be there as well?

It felt more natural to have it this way, though. I mean, you're not flipping through the pages, without going through the list, right?

Going at it right now 😃

**Comment by cristianbote** [comment]
Ok, so, since I have no backend, at all, just fetching on the client, that means, I have to go to the official api, to get the total items, just to display that little counter. Now, if there's another way of getting that total page number, I can easily do that, if not, the performance and numbers will not look good at all.

I'm thinking of keeping this version as it is, without the pagination element, just because it won't work otherwise.

Let me know if there's something else. Cheers!

**Comment by cristianbote** [comment]
So I looked around the apps, I should've done that before, and I think having either the currentPage, e.g. `< prev 1 next >` or have the load more button, but update the url with the current page.

@addyosmani let me know 😄 which one fits the pattern.

**Comment by cristianbote** [comment]
Ok, I added the current page counter. Let me know what you think!

**Comment by cristianbote** [comment]
Friendly ping @addyosmani 😃

**Comment by anubhav7495** [comment]
@cristianbote Nice app man! The performance and WPT scores check out perfectly.

Some minor things before we can merge this app:

1. Pagination should result in URL changes, so that if I refresh while on page `2` or `3`, it wouldn't redirect back to page `1`. **[Important]**
2. On viewports 360px or less wide, the header overflows horizontally.
3. Remove `overflow: auto` property from `section` selector in the stylesheet.
4. Try to use [autoprefixer](https://github.com/postcss/autoprefixer) for safely using flexbox.
5. `<link rel=preload> must have a valid `as` value` i see this logged to console.

Thanks.

**Comment by cristianbote** [comment]
Hey @anubhav7495, thanks for the feedback. Here's the breakdown:

1. ✔️  Done, added `?page=1` in the url as well
2. ❔ This is intended, and it's ugly only on windows browsers which do not emulate a mobile 😃 . Users should be able to swipe through the navigation, on small devices.
3. ✔️  Done. I think I understood what you meant, but let me know if this is not it. You meant to remove the scrollbars from the section and apply it to it's container.
4. ✔️  Done. Thanks for the suggestion!
5. ❔ This is pretty weird, Chrome does complain about it, but it's per specs. Check it out https://w3c.github.io/preload/#link-element-extensions. I have a link header, to preload the json, whish has the `as=fetch` set.

Let me know if there's something else.

Cheers!

**Comment by anubhav7495** [comment]
@cristianbote cool man!

@addyosmani Over to you.

**Comment by addyosmani** [comment]
This all looks excellent, @cristianbote 👌 

Thank you so much for working through the feedback. Looking awesome.

My very very last request and we can land it - can we add the total number of pages we're able to page through?

image

image

**Comment by cristianbote** [comment]
Hey @addyosmani, thanks for looking into it!

Ah, man, that's a deal breaker. I've said this somewhere above, in the comments:

> Ok, so, since I have no backend, at all, just fetching on the client, that means, I have to go to the official api, to get the total items, just to display that little counter. Now, if there's another way of getting that total page number, I can easily do that, if not, the performance and numbers will not look good at all.

Sadly, the performance, would drop significantly. If I remember correctly, from 94 performance score, it would drop to 50 or 60.

I thought every entry would have 500 items in the db, so 17 would always be the last page. But that's not the case, for example https://hacker-news.firebaseio.com/v0/jobstories.json?pretty=print has 24 items, in total.

Not sure what to do.

**Comment by cristianbote** [comment]
Friendly ping @addyosmani

**Comment by anubhav7495** [comment]
@cristianbote Let's do this for a solution.

* Instead of implementing page numbers, just make sure that `next` or `prev` links are disabled when required.

You can check this in practice here: [hnpwa-vue](https://anubhav7495.github.io/hnpwa-vue/#/news)

I use a flag called `hasMore` to implement this whose value is equal to `posts.length ==== 3`

Thanks.

**Comment by cristianbote** [comment]
Hey @anubhav7495, thanks for the suggestion!
I just did that, should be up already.

Thanks for checking it out!
Cheers!

**Comment by anubhav7495** [comment]
@cristianbote cool man.

One minor suggestion from my side which doesn't block the merging of the app:
* Please add `cursor: pointer` property to `.pagination button` in your css.

@addyosmani Over to you.

**Comment by cristianbote** [comment]
@anubhav7495 done! 😉

**Comment by addyosmani** [comment]
@cristianbote @anubhav7495 Thanks for the reviews and working through the feedback! Go team!

Landing this now. I'll adjust the weights (position) to account for recent apps manually.

**Pull request merged**